content/courses/program-security/bump-seed-canonicalization.md

```diff
@@ -1,77 +1,48 @@
-- Because `find_program_address` can be expensive, best practice is to store the
-  derived bump in an account’s data field to be referenced later on when
-  re-deriving the address for verification
-  ```rust
-  #[derive(Accounts)]
-  pub struct VerifyAddress<'info> {
-  	#[account(
-      	seeds = [DATA_PDA_SEED.as_bytes()],
-  	    bump = data.bump
-  	)]
-  	data: Account<'info, Data>,
-  }
-  ```
+- The `create_program_address` function derives a PDA without using the
+  canonical bump, potentially leading to security vulnerabilities.
+- Always use `find_program_address` to ensure the canonical bump is used for PDA
+  derivation.
+- Anchor's `seeds` and `bump` constraints can automatically handle canonical
+  bump usage for PDA initialization and verification.
+- Store the canonical bump in the account's data for efficient re-derivation in
+  future instructions.
 
 ## Lesson
 
-Bump seeds are a number between 0 and 255, inclusive, used to ensure that an
-address derived using
-[`create_program_address`](https://docs.rs/solana-program/latest/solana_program/pubkey/struct.Pubkey.html#method.create_program_address)
-is a valid PDA. The **canonical bump** is the highest bump value that produces a
-valid PDA. The standard in Solana is to _always use the canonical bump_ when
-deriving PDAs, both for security and convenience.
-
-### Insecure PDA derivation using `create_program_address`
-
-Given a set of seeds, the `create_program_address` function will produce a valid
-PDA about 50% of the time. The bump seed is an additional byte added as a seed
-to "bump" the derived address into valid territory. Since there are 256 possible
-bump seeds and the function produces valid PDAs approximately 50% of the time,
-there are many valid bumps for a given set of input seeds.
-
-You can imagine that this could cause confusion for locating accounts when using
-seeds as a way of mapping between known pieces of information to accounts. Using
-the canonical bump as the standard ensures that you can always find the right
-account. More importantly, it avoids security exploits caused by the open-ended
-nature of allowing multiple bumps.
-
-In the example below, the `set_value` instruction uses a `bump` that was passed
-in as instruction data to derive a PDA. The instruction then derives the PDA
-using `create_program_address` function and checks that the `address` matches
-the public key of the `data` account.
-
-```rust
+Bump seeds are numbers between 0 and 255 used to ensure that an address derived
+using `create_program_address` is a valid PDA. The canonical bump is the highest
+valid bump value that produces a valid PDA. For security and consistency, always
+use the canonical bump when deriving PDAs.
+
+### Insecure PDA derivation
+
+The `create_program_address` function produces a valid PDA about 50% of the
+time. Multiple valid bumps can exist for a given set of input seeds, potentially
+causing security issues.
```

Comment:
I want some history here. I personally have never heard of `create_program_address`. I probably got into PDAs after everyone stopped using it.

Find out what the history is here, and tell people. Eg:

> You may find older programs using a function called `create_program_address()` to create PDA addresses...

but not these exact words - whatever the actual truth is. 

Also please tell readers:
 - that `create_program_address` is deprecated (if it is deprecated)
 - why `create_program_address` isn't deprecated (if it is not deprecated)

if it's not deprecated and there's no good reason why, file a  big and list it here.